File: backend/app/schemas/ai.py
"""
backend/app/schemas/ai.py
=========================
Pydantic Schemas for CrimeLens AI Deep Learning API Endpoints.
"""

from pydantic import BaseModel, Field, ConfigDict
from typing import Dict, List, Optional, Any
import uuid
from datetime import datetime, timezone


class CrimePredictionRequest(BaseModel):
    """Payload for real-time crime domain classification."""
    city: str = Field(..., description="City name")
    crime_description: str = Field(..., description="Crime description code")
    victim_age: int = Field(..., ge=0, le=120, description="Victim age")
    victim_gender: str = Field(..., description="Victim gender M/F")
    weapon_used: Optional[str] = Field(default="Unknown", description="Weapon used")
    date_of_occurrence: str = Field(..., description="Date of occurrence dd-mm-yyyy HH:MM")
    time_of_occurrence: str = Field(..., description="Time of occurrence dd-mm-yyyy HH:MM")
    date_reported: Optional[str] = Field(default=None, description="Date reported dd-mm-yyyy HH:MM")
    case_closed: Optional[str] = Field(default="No", description="Whether case is closed")
    crime_code: Optional[int] = Field(default=100, description="Crime code number")

    model_config = ConfigDict(from_attributes=True)


class CrimePredictionResponse(BaseModel):
    """Response payload for crime domain classification."""
    prediction_id: Optional[str] = Field(default=None, description="Prediction UUID")
    predicted_domain: str = Field(..., description="Predicted crime domain")
    domain_code: int = Field(..., description="Domain class index")
    confidence_score: float = Field(..., description="Top-1 confidence score")
    probabilities: Dict[str, float] = Field(..., description="Per-class probabilities")
    model_name: str = Field(..., description="Model class name")
    model_version: str = Field(..., description="Model version string")
    execution_time_ms: int = Field(..., description="Inference latency in ms")
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    model_config = ConfigDict(from_attributes=True)


class ForecastRequest(BaseModel):
    """Payload for multi-horizon sequence forecasting."""
    horizon: int = Field(default=7, description="Forecast horizon in days (7, 30, or 90)")
    recent_sequence: Optional[List[float]] = Field(default=None, description="Optional 30-day historical daily counts")

    model_config = ConfigDict(from_attributes=True)


class ForecastResponse(BaseModel):
    """Response payload for time series forecasting."""
    horizon: int = Field(..., description="Forecast horizon in days")
    forecast_dates: List[str] = Field(..., description="ISO date strings for forecast window")
    predicted_counts: List[float] = Field(..., description="Predicted daily crime counts")
    total_projected_incidents: float = Field(..., description="Sum of predicted counts")
    model_name: str = Field(..., description="Forecaster model class name")
    model_version: str = Field(..., description="Model version string")
    execution_time_ms: int = Field(..., description="Inference latency in ms")

    model_config = ConfigDict(from_attributes=True)


class EmbeddingRequest(BaseModel):
    """Payload for dense vector embedding generation."""
    sample_record: CrimePredictionRequest

    model_config = ConfigDict(from_attributes=True)


class EmbeddingResponse(BaseModel):
    """Response payload for 64-dimensional dense embeddings."""
    embedding_vector: List[float] = Field(..., description="Dense embedding vector")
    embedding_dim: int = Field(default=64, description="Embedding dimension")
    l2_norm: float = Field(default=1.0, description="L2 norm of the embedding")
    model_name: str = Field(default="CrimeEmbeddingNetwork", description="Embedding model name")

    model_config = ConfigDict(from_attributes=True)


class FeatureAttribution(BaseModel):
    feature_name: str
    attribution_score: float


class ExplainabilityRequest(BaseModel):
    """Payload for Captum XAI feature attribution."""
    sample_record: CrimePredictionRequest
    target_class: Optional[int] = Field(default=0, description="Target class index for attribution")

    model_config = ConfigDict(from_attributes=True)


class ExplainabilityResponse(BaseModel):
    """Response payload for feature attribution & sensitivity."""
    predicted_domain: str
    confidence_score: float
    top_contributing_features: List[FeatureAttribution]
    saliency_scores: List[FeatureAttribution]
    model_name: str = Field(default="FTTransformerClassifier")
    method: str = Field(default="Captum Integrated Gradients & Saliency")

    model_config = ConfigDict(from_attributes=True)


class BatchPredictionRequest(BaseModel):
    """Payload for bulk batch inference."""
    records: List[CrimePredictionRequest]

    model_config = ConfigDict(from_attributes=True)


class BatchPredictionResponse(BaseModel):
    """Response payload for bulk batch inference."""
    total_records: int
    predictions: List[CrimePredictionResponse]
    batch_execution_time_ms: int

    model_config = ConfigDict(from_attributes=True)


class ModelHealthResponse(BaseModel):
    """AI Subsystem health probe response."""
    status: str = Field(default="ok", description="Health status string")
    device: str = Field(..., description="Torch device (cpu or cuda)")
    models_loaded: Dict[str, bool] = Field(..., description="Map of model names to load status")
    warmup_complete: bool = Field(default=True)


class ModelMetadataResponse(BaseModel):
    """Metadata response for model version management."""
    version: str = Field(..., description="Model version string")
    framework: str = Field(..., description="Framework version string")
    flagship_model: str = Field(..., description="Primary classifier model name")
    forecaster_model: str = Field(..., description="Forecaster model name")
    embedding_dim: int = Field(default=64, description="Embedding dimension")
    dataset_rows: int = Field(default=40160, description="Training dataset row count")
    dataset_hash_md5: str = Field(..., description="MD5 hash of dataset file")
    saved_at: Optional[str] = None

File: frontend/src/types/ai.ts
/**
 * TypeScript interfaces for CrimeLens AI – Phase 8
 * Mirrors backend app/schemas/ai.py and app/schemas/mlops.py exactly.
 */

// ─── AI Prediction ─────────────────────────────────────────────────────────

export interface CrimePredictionRequest {
  city: string;
  crime_description: string;
  victim_age: number;
  victim_gender: string;
  weapon_used?: string;
  date_of_occurrence: string;   // dd-mm-yyyy HH:MM
  time_of_occurrence: string;   // dd-mm-yyyy HH:MM
  date_reported?: string | null;
  case_closed?: string;         // "Yes" | "No"
  crime_code?: number;
}

export interface CrimePredictionResponse {
  prediction_id: string | null;
  predicted_domain: string;
  domain_code: number;
  confidence_score: number;
  probabilities: Record<string, number>;
  model_name: string;
  model_version: string;
  execution_time_ms: number;
  timestamp: string;
}

// ─── Forecasting ────────────────────────────────────────────────────────────

export interface ForecastRequest {
  horizon: 7 | 30 | 90;
  recent_sequence?: number[] | null;
}

export interface ForecastResponse {
  horizon: number;
  forecast_dates: string[];
  predicted_counts: number[];
  total_projected_incidents: number;
  model_name: string;
  model_version: string;
  execution_time_ms: number;
}

// ─── Embedding ──────────────────────────────────────────────────────────────

export interface EmbeddingResponse {
  embedding_vector: number[];
  embedding_dim: number;
  l2_norm: number;
  model_name: string;
}

// ─── Explainability (XAI) ───────────────────────────────────────────────────

export interface FeatureAttribution {
  feature_name: string;
  attribution_score: number;
}

export interface ExplainabilityRequest {
  sample_record: CrimePredictionRequest;
  target_class?: number;
}

export interface ExplainabilityResponse {
  predicted_domain: string;
  confidence_score: number;
  top_contributing_features: FeatureAttribution[];
  saliency_scores: FeatureAttribution[];
  model_name: string;
  method: string;
}

// ─── Batch Prediction ───────────────────────────────────────────────────────

export interface BatchPredictionResponse {
  total_records: number;
  predictions: CrimePredictionResponse[];
  batch_execution_time_ms: number;
}

// ─── Model Health & Metadata ────────────────────────────────────────────────

export interface ModelHealthResponse {
  status: string;
  device: string;
  models_loaded: Record<string, boolean>;
  warmup_complete: boolean;
}

export interface ModelMetadataResponse {
  version: string;
  framework: string;
  flagship_model: string;
  forecaster_model: string;
  embedding_dim: number;
  dataset_rows: number;
  dataset_hash_md5: string;
  saved_at?: string | null;
}

// ─── MLOps – Model Registry ─────────────────────────────────────────────────

export interface ModelVersionInfo {
  model_name: string;
  version: string;
  training_date?: string | null;
  dataset_hash: string;
  dataset_version?: string | null;
  framework_version: string;
  accuracy_metrics: Record<string, number>;
  metrics: Record<string, unknown>;
  status: 'Active' | 'Archived';
  artifact_path?: string | null;
}

export interface ModelListResponse {
  models: ModelVersionInfo[];
  count: number;
}

// ─── MLOps – Analytics ──────────────────────────────────────────────────────

export interface AnalyticsSummary {
  total_predictions: number;
  predictions_today: number;
  predictions_this_month: number;
  average_confidence: number;
  average_inference_latency_ms: number;
  most_predicted_crime_domain?: string | null;
  human_review_count: number;
  low_confidence_count: number;
}

export interface AnalyticsResponse {
  summary: AnalyticsSummary;
  prediction_distribution: Record<string, number>;
  model_usage: Record<string, unknown>[];
  top_confidence_predictions: Record<string, unknown>[];
  low_confidence_predictions: Record<string, unknown>[];
  daily_prediction_trend: Record<string, unknown>[];
  monthly_prediction_trend: Record<string, unknown>[];
  charts: Record<string, unknown>;
}

// ─── MLOps – Drift ──────────────────────────────────────────────────────────

export type DriftLevel = 'LOW' | 'MEDIUM' | 'HIGH';

export interface DriftComponent {
  score: number;
  level: DriftLevel;
  affected_features: string[];
}

export interface DriftStatusResponse {
  drift_score: number;
  overall_level: DriftLevel;
  feature_drift: DriftComponent;
  confidence_drift: Record<string, unknown>;
  prediction_distribution_drift: Record<string, unknown>;
  sample_size: number;
  warnings: string[];
  timestamp: string;
}

export interface DriftReportResponse {
  drift_score: number;
  drift_level: DriftLevel;
  drift_type: string[];
  affected_features: string[];
  recommendation: string;
  warnings: string[];
  timestamp: string;
  details: Record<string, unknown>;
}

// ─── MLOps – System Status ──────────────────────────────────────────────────

export interface SystemStatusResponse {
  api_status: string;
  database_status: string;
  database_detail: string;
  supabase_status: string;
  supabase_detail: string;
  model_status: string;
  model_errors: string[];
  registry_status: string;
  registry_detail: string;
  active_version?: string | null;
  memory_usage_mb?: number | null;
  cpu_usage_percent?: number | null;
  uptime_seconds: number;
  project_root: string;
}

// ─── Derived / UI helpers ───────────────────────────────────────────────────

export type RiskLevel = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';

export function confidenceToRisk(confidence: number): RiskLevel {
  if (confidence >= 0.85) return 'CRITICAL';
  if (confidence >= 0.65) return 'HIGH';
  if (confidence >= 0.40) return 'MEDIUM';
  return 'LOW';
}

export function formatConfidence(score: number): string {
  return `${(score * 100).toFixed(1)}%`;
}

export function formatUptime(seconds: number): string {
  const h = Math.floor(seconds / 3600);
  const m = Math.floor((seconds % 3600) / 60);
  return h > 0 ? `${h}h ${m}m` : `${m}m`;
}
